gpflow/likelihoods/base.py

```diff
@@ -286,7 +286,71 @@ def _variational_expectations(self, Fmu, Fvar, Y):
         raise NotImplementedError
 
 
-class ScalarLikelihood(Likelihood):
+class QuadratureLikelihood(Likelihood):
+    @property
+    def quadrature(self):
+        raise NotImplementedError()
+
+    def _predict_mean_and_var(self, Fmu, Fvar):
+        r"""
+        :param Fmu: mean function evaluation Tensor, with shape [..., latent_dim]
+        :param Fvar: variance of function evaluation Tensor, with shape [..., latent_dim]
+        :returns: mean and variance, both with shape [..., observation_dim]
+        """
+
+        def conditional_y_squared(F):
+            return self.conditional_variance(F) + self.conditional_mean(F) ** 2
+
+        integrands = [self.conditional_mean, conditional_y_squared]
+        E_y, E_y2 = self.quadrature(integrands, Fmu, Fvar)
+        V_y = E_y2 - E_y ** 2
+        return E_y, V_y
+
+    def _quadrature_log_prob(self, F, Y):
+        return tf.expand_dims(self.log_prob(F, Y), -1)
+
+    def _predict_log_density(self, Fmu, Fvar, Y):
+        r"""
+        :param Fmu: mean function evaluation Tensor, with shape [..., latent_dim]
+        :param Fvar: variance of function evaluation Tensor, with shape [..., latent_dim]
+        :param Y: observation Tensor, with shape [..., observation_dim]:
+        :returns: variational expectations, with shape [...]
+        """
+        return tf.squeeze(self.quadrature.logspace(self._quadrature_log_prob, Fmu, Fvar, Y), -1)
+
+    def _variational_expectations(self, Fmu, Fvar, Y):
+        r"""
+        :param Fmu: mean function evaluation Tensor, with shape [..., latent_dim]
+        :param Fvar: variance of function evaluation Tensor, with shape [..., latent_dim]
+        :param Y: observation Tensor, with shape [..., observation_dim]:
+        :returns: log predictive density, with shape [...]
+        """
+        return tf.squeeze(self.quadrature(self._quadrature_log_prob, Fmu, Fvar, Y), -1)
+
+
+class NDiagGHQuadratureLikelihood(QuadratureLikelihood):
+    def __init__(self, n_gh: int = 20, **kwargs):
+        super().__init__(**kwargs)
+        self.n_gh = n_gh
+        self._quadrature = None
+
+    @property
+    def quadrature(self):
+        if self._quadrature is None:
+            if self.latent_dim is None:
+                raise Exception(
```

Comment:
Should/could this be a more specific exception? Maybe a custom one?

Reply:
Do you mean implementing a custom gpflow error? I think it would be a better choice, indeed. To be honest, I've never done something like this, some guidance would be appreciated :).

Reply:
In this PR I wouldn't introduce custom exceptions as AFAIK we don't have them anywhere else in the gpflow code base. If we want custom exceptions we should probably discuss this more generally and create a separate PR for it.

Reply:
Resolving this one and saving the discussion of custom exceptions for another PR!